// @ion-phaser/core/dist/esm/ion-phaser.entry.js
import { r as registerInstance, g as getElement } from './core-58ca22c8.js';

const IonPhaser = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        /**
         * To initialize the plugin manually
         */
        this.initialize = true;
        this.initializeGame = (game = this.game) => {
            if (!game)
                return;
            if (game.instance) {
                throw new Error("A Phaser game already exist");
            }
            game.parent = game.parent || this.el;
            game.instance = new Phaser.Game(game);
        };
    }
    onGameChange(game) {
        if (this.initialize && !this.getGameInstance()) {
            this.initializeGame(game);
        }
    }
    onInitialize(initialize) {
        if (initialize && !this.getGameInstance()) {
            this.initializeGame();
        }
    }
    /**
     * Get the Phaser game instance
     */
    async getInstance() {
        return Promise.resolve(this.getGameInstance());
    }
    /**
     * Destroy the Phaser game instance
     */
    async destroy() {
        if (this.getGameInstance()) {
            this.game.instance.destroy(true);
            this.game.instance = null;
        }
    }
    componentWillLoad() {
        if (!this.getGameInstance() && this.initialize) {
            this.initializeGame();
        }
    }
    componentDidUnload() {
        this.destroy();
    }
    getGameInstance() {
        return this.game && this.game.instance;
    }
    get el() { return getElement(this); }
    static get watchers() { return {
        "game": ["onGameChange"],
        "initialize": ["onInitialize"]
    }; }
    static get style() { return "ion-phaser{display:block}"; }
};

export { IonPhaser as ion_phaser };
