// @ion-phaser/core/dist/collection/components/ion-phaser/ion-phaser.js
export class IonPhaser {
    constructor() {
        /**
         * To initialize the plugin manually
         */
        this.initialize = true;
        this.initializeGame = (game = this.game) => {
            if (!game)
                return;
            if (game.instance) {
                throw new Error("A Phaser game already exist");
            }
            game.parent = game.parent || this.el;
            game.instance = new Phaser.Game(game);
        };
    }
    onGameChange(game) {
        if (this.initialize && !this.getGameInstance()) {
            this.initializeGame(game);
        }
    }
    onInitialize(initialize) {
        if (initialize && !this.getGameInstance()) {
            this.initializeGame();
        }
    }
    /**
     * Get the Phaser game instance
     */
    async getInstance() {
        return Promise.resolve(this.getGameInstance());
    }
    /**
     * Destroy the Phaser game instance
     */
    async destroy() {
        if (this.getGameInstance()) {
            this.game.instance.destroy(true);
            this.game.instance = null;
        }
    }
    componentWillLoad() {
        if (!this.getGameInstance() && this.initialize) {
            this.initializeGame();
        }
    }
    componentDidUnload() {
        this.destroy();
    }
    getGameInstance() {
        return this.game && this.game.instance;
    }
    static get is() { return "ion-phaser"; }
    static get originalStyleUrls() { return {
        "$": ["ion-phaser.css"]
    }; }
    static get styleUrls() { return {
        "$": ["ion-phaser.css"]
    }; }
    static get properties() { return {
        "game": {
            "type": "unknown",
            "mutable": true,
            "complexType": {
                "original": "GameInstance",
                "resolved": "GameInstance",
                "references": {
                    "GameInstance": {
                        "location": "import",
                        "path": "../models"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The configuration of the game"
            }
        },
        "initialize": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "To initialize the plugin manually"
            },
            "attribute": "initialize",
            "reflect": false,
            "defaultValue": "true"
        }
    }; }
    static get methods() { return {
        "getInstance": {
            "complexType": {
                "signature": "() => Promise<Game>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    },
                    "Game": {
                        "location": "import",
                        "path": "phaser"
                    }
                },
                "return": "Promise<Game>"
            },
            "docs": {
                "text": "Get the Phaser game instance",
                "tags": []
            }
        },
        "destroy": {
            "complexType": {
                "signature": "() => Promise<void>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<void>"
            },
            "docs": {
                "text": "Destroy the Phaser game instance",
                "tags": []
            }
        }
    }; }
    static get elementRef() { return "el"; }
    static get watchers() { return [{
            "propName": "game",
            "methodName": "onGameChange"
        }, {
            "propName": "initialize",
            "methodName": "onInitialize"
        }]; }
}
